experimental/parser/parse_delimited.go

```diff
@@ -100,7 +156,13 @@ func (d delimited[T]) iter(yield func(value T, delim token.Token) bool) {
 			}).Apply(report.Snippetf(delim, "first delimiter is here"))
 		}
 
-		if !yield(v, delim) || (d.required && delim.Nil()) {
+		if !yield(v, delim) {
+			break
+		}
+
+		if delim.Nil() && d.required && !d.exhaust {
+			// In non-exhaust mode, if we miss a required comma just bail.
+			// Otherwise, go again to parse another thing.
```

Comment:
I think "non-exhaust mode" is for ranges. But I think we could handle missing commas in that case, too, if we also have some sort of "stop" marker -- so we stop if/when we see a semicolon and can otherwise continue past missing commas.